### PR TITLE
Added a way to sort comments on an article 

### PR DESCRIPTION
**Pull request by vishaldeepak**
## What type of PR is this? (check all applicable)

- [*] Feature

## Description

Add a way to sort comments. Comments can be sorted in three ways
- latest
- oldest
- score

Currently deals Point 3 of the issue 
User, post-level

## Related Tickets & Documents
Deals with #16133 

## QA Instructions, Screenshots, Recordings
How it looks :
Screenshot 2022-02-25 at 12 20 02 PM

### UI accessibility concerns?

The view was followed using [this comment](https://github.com/forem/forem/issues/16133#issue-1105555301)
I require svg files for the icons

Want to add accessibility. But would like to confirm the view

Currently
- Selectable via keyboard
- Didnt get any new errors when running Axe tool

## Added/updated tests?

- [ ] No, and this is why: Would like to verify approach first , then add and fix existing cases

**Update**
- Added test cases

## [Forem core team only] How will this change be communicated?

_Will this PR introduce a change that impacts Forem members or creators, the
development process, or any of our internal teams? If so, please note how you
will share this change with the people who need to know about it._

- [ ] I've updated the [Developer Docs](https://developers.forem.com) or
      [Storybook](https://storybook.forem.com/) (for Crayons components)
- [ ] This PR changes the Forem platform and our documentation needs to be
      updated. I have filled out the
      [Changes Requested](https://github.com/forem/admin-docs/issues/new?assignees=&labels=&template=changes_requested.md)
      issue template so Community Success can help update the Admin Docs
      appropriately.
- [ ] I've updated the README or added inline documentation
- [ ] I've added an entry to
      [`CHANGELOG.md`](https://github.com/forem/forem/tree/main/CHANGELOG.md)
- [ ] I will share this change in a [Changelog](https://forem.dev/t/changelog)
      or in a [forem.dev](http://forem.dev) post
- [ ] I will share this change internally with the appropriate teams
- [ ] I'm not sure how best to communicate this change and need help
- [ ] This change does not need to be communicated, and this is why not: _please
      replace this line with details on why this change doesn't need to be
      shared_

## [optional] What gif best describes this PR or how it makes you feel?

alt_text

### CODE REVIEW

File: app/controllers/stories_controller.rb
Comment (vishaldeepak) [reply app/controllers/stories_controller.rb:419]: @citizen428 fixed

File: app/assets/stylesheets/views/comments.scss
Comment (nickytonline) [inline app/assets/stylesheets/views/comments.scss:283]: Refactor (blocking): Styling for a dropdown is not required. Use the CSS class `crayons-dropdown`. Using this class wires up a bunch of functionality to the dropdown to make it more accessible as well.

For margins and padding stick to our CSS variables, e.g. ✅  `padding: var(--su-4);` ❌  `margin-top: 0.25rem;`.

@ludwiczakpawel, do you have any design input here?

File: app/views/articles/_full_comment_area.html.erb
Comment (nickytonline) [inline app/views/articles/_full_comment_area.html.erb:5]: Refactor (blocking): A header cannot be used as the trigger for the dropdown as it causes accessibility issues. A sort button to the right of the header as the trigger will be more accessible. cc: @aitchiss

File: app/views/articles/_full_comment_area.html.erb
Comment (nickytonline) [inline app/views/articles/_full_comment_area.html.erb:11]: Refactor (blocking): There are missing `<%=` for rendering partials.

```suggestion
<%= render "comments/sort_option",
```

CleanShot 2022-03-07 at 10 27 22

**Review by nickytonline — CHANGES_REQUESTED** [review]
@vishaldeepak, thanks so much for the PR. A dropdown is a right approach, but there are some issues with the implementation. I've left some feedback.

File: app/views/articles/_full_comment_area.html.erb
Comment (vishaldeepak) [reply app/views/articles/_full_comment_area.html.erb:11]: thats really odd, must be some auto-fix while committing. Will fix this

File: app/views/articles/_full_comment_area.html.erb
Comment (aitchiss) [reply app/views/articles/_full_comment_area.html.erb:5]: Checking out the designs on https://github.com/forem/forem/issues/16133, I think you can use an [icon only button](https://storybook.forem.com/?path=/story/components-buttons--icon-only) with the `expand.svg` icon.

In terms of accessibility, be sure to give the button an `aria-label` that describes the function ("sort comments" would work). We also have a helper to initialize dropdown expand/collapse behaviours in JS (incuding accessibility features) - check out https://github.com/forem/forem/blob/main/app/javascript/utilities/dropdownUtils.js#L118

File: app/assets/stylesheets/views/comments.scss
Comment (ludwiczakpawel) [reply app/assets/stylesheets/views/comments.scss]: I know im late to the party, but it seems like it has been fixed...?

Comment (vishaldeepak) [reply app/assets/stylesheets/views/comments.scss]: yes, I was waiting on test cases to inform you guys

**Comment by vishaldeepak** [comment]
@nickytonline @ludwiczakpawel  Request changes made , and added relevant test cases added. 
I was unable to start cypress tests locally. Ran into errors. The [contributing docs](https://developers.forem.com/contributing-guide/forem/) don't mention how to run this tests I suppose. I fixed existing test cases for cypress , since that was just a string change, though was unable to add a new tests on them.

File: app/views/comments/_sort_option.html.erb
Comment (aitchiss) [reply app/views/comments/_sort_option.html.erb:1]: > would a `<nav />` with an `aria-label` (to distinguish it from other `<nav />`s) that contains a list of hyperlinks make more sense?

Yes, I think in this case that's most semantic/accessible. Not all lists of links _need_ to be in a `nav`, but I think in this case it could be helpful since they are a very distinct group. So what we want is something like:

```html
<nav aria-labelledby="sort-comments-heading">
  <h3 id="sort-comments-heading">Sort discussion:</h3>
  <ul>
    <li>
      <a href="<link to the sort option>" aria-describedby="<sort option name>-description-text">sort option text</a>
      <span id="<sort option name>-description-text">description text for sort option</span>
    </li>
  <ul>
</nav>
```
This is just pseudo-code, but essentially we want a list of links inside a `nav`, and we want all descriptions to be linked to the link itself with `aria-describedby`. `Top`, `Latest`, etc should not be headings.

Also, `aria-current="page"` should be added to the `<a>` tag when that item is the currently selected sort option (we could even control the appearance of the checkmark icon in CSS with the `[aria-current="page"]` selector)

Let me know if I can provide any more clarification on this!

Comment (vishaldeepak) [reply app/views/comments/_sort_option.html.erb:1]: @aitchiss Are you implying that we also load the svg from CSS? 

How would you set color to the svg via css. For example I tried some code 

HTML
```
<li class="comment-sort-option">
  <a href="/admin_mcadmin/this-is-title-2m9f?comments_sort=top" class="comment-sort-option__header first-nav-link" aria-describedby="top-description-text" aria-current="page">
    Top
  </a>
  <span id="top-description-text" class="crayons-field__description">Most upvoted and relevant comments will be first</span>
</li>
```

Css
```
    [aria-current="page"]::before{
      content: '';
      background-image: url(image_path('checkmark.svg'));
      position: relative;
      display: block;
      background-size: 28px 28px;
      height: 28px;
      width: 28px;
    }
```
The problem is that the svg loads but not in the font-color we want.  There's a way I found on [stack](https://stackoverflow.com/a/42317014/3480449).  Should we go that route? 

Other way would be to load the svg in html, but we might end up in code that might be undesirable
```
<li class="comment-sort-option">
  <span class="sort-selected">
      <svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24" fill="none" class="crayons-icon" role="img" aria-labelledby="abks7e4kr1ixjsoqqloov97nkecs1rbl"><title id="abks7e4kr1ixjsoqqloov97nkecs1rbl">Selected Sort Option</title>
    <path d="M10 15.172l9.192-9.193 1.415 1.414L10 18l-6.364-6.364 1.414-1.414 4.95 4.95z"></path>
</svg>

  </span>
  <span>
    <a href="/admin_mcadmin/this-is-title-2m9f?comments_sort=top" class="comment-sort-option__header first-nav-link" aria-describedby="top-description-text" aria-current="page">
      Top
    </a>
    <span id="top-description-text" class="crayons-field__description">Most upvoted and relevant comments will be first</span>
  </span>
</li>
```

Comment (aitchiss) [reply app/views/comments/_sort_option.html.erb:1]: Yes sorry I wasn't very clear in my suggestion! We can load the svg in the `html.erb` file, using the `crayons_icon_tag` helper function, adding it inside the link, and including a class name like `sort-selected-icon`, keeping it hidden by default (using the helper class name 'hidden'). We could then reveal it using something like

```scss
[aria-current="page"] {
  .sort-selected {
    display: inline-block;
  }
}
```

Comment (aitchiss) [reply app/views/comments/_sort_option.html.erb:1]: (but if showing/hiding the icon in this way adds too much complexity, I don't mind going another direction - it was just an idea! 🙂 )

Comment (vishaldeepak) [reply app/views/comments/_sort_option.html.erb]: changes made.

File: app/assets/stylesheets/views/comments.scss
Comment (aitchiss) [inline app/assets/stylesheets/views/comments.scss:284]: (UI bug / blocking) - This applies an elevated z-index to the whole area, including the heading, when we only want it to apply to the actual dropdown. The result is that it's possible to scroll the heading into the top nav bar:

heading present in top nav bar

You should be able to rely on the `crayons-dropdown` class on the dropdown content itself to control `z-index`, but let me know if you have any problems with that

File: app/assets/stylesheets/views/comments.scss
Comment (aitchiss) [inline app/assets/stylesheets/views/comments.scss:295]: refactor (blocking) - I think almost all of these you don't need, as they're applied by the `crayons-dropdown` class. The only one I think you require is the padding, and you can add this with a helper classname directly on the HTML element - `p-4`

File: app/assets/stylesheets/views/comments.scss
Comment (aitchiss) [inline app/assets/stylesheets/views/comments.scss:299]: refactor (not-blocking) - prefer the helper utility classes `mt-2 pl-7` on the HTML element

File: app/assets/stylesheets/views/comments.scss
Comment (aitchiss) [inline app/assets/stylesheets/views/comments.scss:304]: refactor (blocking) - I don't think this line is needed

File: app/assets/stylesheets/views/comments.scss
Comment (aitchiss) [inline app/assets/stylesheets/views/comments.scss:310]: refactor (blocking) - this class doesn't seem to be used anywhere, so let's delete it 
```suggestion
```

File: app/assets/stylesheets/views/comments.scss
Comment (aitchiss) [inline app/assets/stylesheets/views/comments.scss:314]: refactor (blocking) - I think you went a different route with this in the end, and the icon is only added in the HTML next to the currently selected option. So we no longer need this CSS property, or the related `aria-current` selector

```suggestion
```

File: app/assets/stylesheets/views/comments.scss
Comment (aitchiss) [inline app/assets/stylesheets/views/comments.scss:325]:
```suggestion
```

(as above)

File: app/assets/stylesheets/views/comments.scss
Comment (aitchiss) [inline app/assets/stylesheets/views/comments.scss:332]: refactor (blocking) - as there's only one property here, let's favour using helper utility classes on the element itself. You should be able to use `mb-3` classname on the element to get the same effect

**Review by aitchiss — CHANGES_REQUESTED** [review]
Thanks for your work on this one! It's working really well, but I've requested a few refactoring/cleanup changes that will help make this more robust and maintainable in future.

I've also asked that we add some Cypress tests for this flow, but if you need assistance with that, just let me know and either myself or someone else from the team can chip in with those!

File: app/assets/stylesheets/views/comments.scss
Comment (vishaldeepak) [reply app/assets/stylesheets/views/comments.scss:284]: That was my initial thought as well. But it seemed to cause a UI issue. 
Screenshot 2022-03-26 at 2 00 10 PM

File: app/assets/stylesheets/views/comments.scss
Comment (vishaldeepak) [reply app/assets/stylesheets/views/comments.scss:295]: good catch. But without the width property, the end result is a bit different from what was posted in the issue.
Screenshot 2022-03-26 at 2 10 38 PM

File: spec/system/articles/user_visits_an_article_spec.rb
Comment (vishaldeepak) [reply spec/system/articles/user_visits_an_article_spec.rb:188]: Hey I need  help to setup cypress on local. I was unable to start cypress tests locally. Ran into errors. The [contributing docs](https://developers.forem.com/contributing-guide/forem/) don't mention how to run this tests I suppose. 
I tried starting the server locally via `rake cypress:open`
But I run into an error
Screenshot 2022-03-26 at 11 25 19 PM

Comment (aitchiss) [reply spec/system/articles/user_visits_an_article_spec.rb:188]: Ah can you try running the specs using the command `yarn e2e`. You'll see a message in the console asking you if you want to set up your database before running, and please type 'Y' to confirm that yep we want this! Hopefully that should resolve your issues.

[The E2E testing docs for Forem are here](https://developers.forem.com/tests/e2e-tests) - let me know if you notice things missing from them though, as we're always working on improving these

File: app/assets/stylesheets/views/comments.scss
Comment (vishaldeepak) [reply app/assets/stylesheets/views/comments.scss:284]: @aitchiss do let me know if you figure this one out

File: app/javascript/packs/commentDropdowns.js
Comment (aitchiss) [inline app/javascript/packs/commentDropdowns.js:2]: This file is related to the top member navigaton, and isn't something we need to use here - trying to repurpose this is part of what is causing the stacking issues

Instead, see below, and comments in `app/views/articles/_full_comment_area.html.erb` for how we can initialize the dropdown

File: app/javascript/packs/commentDropdowns.js
Comment (aitchiss) [inline app/javascript/packs/commentDropdowns.js:88]:
```suggestion
  const sortCommentsDropdownTrigger = document.getElementById(
    'toggle-comments-sort-dropdown',
  );

  if (!sortCommentsDropdownTrigger.dataset.initialized) {
     initializeDropdown({
       triggerElementId: 'toggle-comments-sort-dropdown',
       dropdownContentId: 'comments-sort-dropdown-container',
     });
    sortCommentsDropdownTrigger.dataset.initialized = true
  }
```

This refactors to use the `initializeDropdown` helper used elsewhere. This helper is designed to work with `crayons-dropdown` classes 🙂

File: app/views/articles/_full_comment_area.html.erb
Comment (aitchiss) [inline app/views/articles/_full_comment_area.html.erb:38]:
```suggestion
<div class="flex">
          <h2>
            <%= t("views.articles.comments.subtitle.#{@comments_order}_html",
                num: tag.span(t("views.articles.comments.num", num: @article.comments_count), class: "js-comments-count", data: { comments_count: @article.comments_count })) %>
          </h2>
          <button id="toggle-comments-sort-dropdown" class="crayons-btn crayons-btn--ghost crayons-btn--icon-left" id="toggle-comments-sort-dropdown" aria-controls="comments-sort-dropdown-container" aria-expanded="false" aria-label="Sort comments" aria-haspopup="true">
            <%= inline_svg_tag("expand.svg", aria_hidden: true, title: t("views.moderations.actions.admin.icon")) %>
          </button>
        </div>
        <nav id="comments-sort-dropdown-container" class="crayons-dropdown p-4" aria-labelledby="sort-comments-heading">
          <h4 class="comments-sort-title">Sort discussion: </h4>
          <ul class="comments-sort-dropdown__list">
            <%= render "comments/sort_option",
                        sort_order: "top",
                        sort_title: "Top",
                        sort_description: "Most upvoted and relevant comments will be first",
                        show_selected: @comments_order == "top",
                        html_classes: "first-nav-link" %>
            <%= render "comments/sort_option",
                        sort_order: "latest",
                        sort_title: "Latest",
                        sort_description: "Most recent comments will be first",
                        show_selected: @comments_order == "latest",
                        html_classes: "second-last-nav-link" %>
            <%= render "comments/sort_option",
                        sort_order: "oldest",
                        sort_title: "Oldest",
                        sort_description: "The oldest comments will be first",
                        show_selected: @comments_order == "oldest",
                        html_classes: "last-nav-link" %>
          </ul>
        </nav>
```

Here I have again refactored to remove any conflation with the member top navigation menu, as mentioned in the `commentDropdowns` file. Removing the conflation allows the dropdown to be opened/closed, relying on the `crayons-dropdown` class and without any strange stacking issues (it is expected that the open dropdown would appear over the sticky header when scrolled):

https://user-images.githubusercontent.com/20773163/160593995-3315e4bb-82b6-44e7-9ce9-5b630ecdef6b.mp4

**Review by aitchiss — COMMENTED** [review]
Added some comments on how tI managed to resolve the stacking issues locally. It seems the issue is trying to apply the top member nav menu classes/code to this dropdown, which as I mentioned before, isn't what we want to do 🙂 

Let me know if you have any more difficulty!

File: spec/system/articles/user_visits_an_article_spec.rb
Comment (vishaldeepak) [reply spec/system/articles/user_visits_an_article_spec.rb:188]: added tests in cypress, let me know If should remove the the above test?

File: app/javascript/packs/commentDropdowns.js
Comment (vishaldeepak) [reply app/javascript/packs/commentDropdowns.js:88]: thanks so much, I made another change that was a bit easier

**Comment by vishaldeepak** [comment]
@aitchiss  Thanks so much for your help. I have added cypress test cases and cleand up code. Should I remove `system/articles` spec

File: app/javascript/packs/commentDropdowns.js
Comment (aitchiss) [reply app/javascript/packs/commentDropdowns.js:24]: perfect! should we maybe keep the new query selector in line with the existing ones, e.g.
```suggestion
    'button[id^=comment-dropdown-trigger], button[id^=comment-profile-preview-trigger-], button[id^=toggle-comments-sort-dropdown],
```

(this isn't in any way blocking, just a thought 😄 )

File: cypress/integration/seededFlows/articleFlows/sortCommentsOnArticle.spec.js
Comment (aitchiss) [inline cypress/integration/seededFlows/articleFlows/sortCommentsOnArticle.spec.js:80]: (same as above)

File: spec/system/articles/user_visits_an_article_spec.rb
Comment (aitchiss) [reply spec/system/articles/user_visits_an_article_spec.rb:188]: Yeah I think we can remove this one now, as it's well covered in Cypress 🔥

**Review by aitchiss — COMMENTED** [review]
Nice work adding the Cypress tests - they look great! I'll take this for another little spin now 😄

**Comment by aitchiss** [comment]
I noticed a bit of a strange bug, where once I select an item from the dropdown, the page refreshes with the right content, but then next time I click the dropdown button, nothing happens:

https://user-images.githubusercontent.com/20773163/161053408-22e8d1b8-c4a6-4170-8180-263b85a7c6c1.mp4


I had a little investigate and it seems to be this code in `dropdownUtils` causing the issue:

https://github.com/forem/forem/blob/0403e78f08b2de9e90160be721896ab9879c08b9/app/javascript/utilities/dropdownUtils.js#L167-L184

It seems like what it thinks is the `triggerButton` is still the button on the _last_ page, rather than the current one 🙈  I think this is less a bug with your code, and more a bug with the utility function. How would you feel about adding the fix in this PR though? I think all we need to do is to update `clickOutsideListener` to get a fresh handle of the button each time, which would be a bit more robust e.g. `const triggerButton = document.getElementById(triggerElementId);`

**Comment by vishaldeepak** [comment]
> I noticed a bit of a strange bug, where once I select an item from the dropdown, the page refreshes with the right content, but then next time I click the dropdown button, nothing happens:
> CleanShot.2022-03-31.at.13.20.49.mp4
> 
> I had a little investigate and it seems to be this code in `dropdownUtils` causing the issue:
> 
> https://github.com/forem/forem/blob/0403e78f08b2de9e90160be721896ab9879c08b9/app/javascript/utilities/dropdownUtils.js#L167-L184
> 
> It seems like what it thinks is the `triggerButton` is still the button on the _last_ page, rather than the current one 🙈 I think this is less a bug with your code, and more a bug with the utility function. How would you feel about adding the fix in this PR though? I think all we need to do is to update `clickOutsideListener` to get a fresh handle of the button each time, which would be a bit more robust e.g. `const triggerButton = document.getElementById(triggerElementId);`

**Comment by vishaldeepak** [comment]
@aitchiss 
hey, made changes accordingly. Changing the `clickOustideListener` function, caused a Jest test to fail. `CommentSubscription.test.jsx:62:12`. I had to add another conditional in `clickOustideListener` to make this work.

**Review by aitchiss — APPROVED** [review]
Great work @vishaldeepak! Thank you for making those changes - it's working great now 🔥 

We'll get one more review on this and then we should be good to go 🚀

File: app/controllers/stories_controller.rb
Comment (djuber) [inline app/controllers/stories_controller.rb:419]: refactor, non-blocking: the logic here about which sort options are valid, and the "top" default sort option if not provided, might be moved to a method in the Comment model, passing the `:comment_sort` param to the method and receiving a string back. That keeps the http request details (params provided) in the controller, but the rules about sorting behavior in the model (where you added the constant).

The code smell here for me is reaching into the value of the class's constant, outside of a controller tied to that model like CommentsController.

Something like this might make sense. 

```ruby
class Comment
  def self.sort_order(option)
     # code from fetch_sort_order here    
  end
end
```

I'll merge this as-is, if this logic needs to change in the future we can revisit moving the decision making out of the controller.

**Review by djuber — APPROVED** [review]

**Pull request merged**
